File: NeetCode150/two_pointers/1_valid_palindrome.cc
class Solution {
    public:
        // Time: O(n)
        // Space: O(1)
        bool isPalindrome(string s) {
            int beg = 0;
            int end = s.size() - 1;
    
            while (beg <= end) {
                if (isalnum(s[beg]) && isalnum(s[end])) {
                    if (tolower(s[beg]) != tolower(s[end])) {
                        return false;
                    }
                    beg++;
                    end--;
                } else if (isalnum(s[beg]) && !isalnum(s[end])) {
                    end--;
                } else if (!isalnum(s[beg]) && isalnum(s[end])) {
                    beg++;
                } else {
                    beg++;
                    end--;
                }
            }
    
            return true;
        }
    };

File: NeetcodeBlind75/two_pointers/1_valid_palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        """
        Algorithm intuition:
            - we can use a two pointer approach
            - start with left and right index pointer
            - check if s[left] == s[right]
                - if at any time false then return False
            - if element at index is not alphanumber we increase just the left or right pointer accordingly
        Time: O(n)
        Space: O(1)
        """
        left, right = 0, len(s) - 1

        while left <= right:
            if not s[left].isalnum():
                left += 1
            elif not s[right].isalnum():
                right -= 1
            elif s[left].lower() == s[right].lower():
                left += 1
                right -= 1
            else:
                return False

        return True
